lib/membrane_mp4/demuxer/isom.ex

```diff
@@ -126,28 +199,141 @@ defmodule Membrane.MP4.Demuxer.ISOM do
     {[], %{state | samples_info: samples_info, partial: rest}}
   end
 
-  def handle_process(:input, buffer, ctx, %{samples_info: nil} = state) do
-    # Parse the boxes we have received
-    {boxes, rest} = Container.parse!(state.partial <> buffer.payload)
-    boxes = state.boxes ++ boxes
+  def handle_process(:input, buffer, ctx, state) do
+    {new_boxes, rest} = Container.parse!(state.partial <> buffer.payload)
+
+    state = %{
+      state
+      | boxes: state.boxes ++ new_boxes,
+        boxes_size:
+          state.boxes_size + byte_size(state.partial <> buffer.payload) - byte_size(rest)
+    }
+
+    maybe_header = parse_header(rest)
 
-    parsing_mdat? =
-      case parse_header(rest) do
-        nil -> false
-        header -> header.name == :mdat
+    started_parsing_mdat? =
+      cond do
+        :mdat in Keyword.keys(state.boxes) -> true
+        maybe_header != nil -> maybe_header.name == :mdat
+        true -> false
       end
 
-    state = %{state | boxes: boxes, partial: rest, parsing_mdat?: parsing_mdat?}
+    state = %{state | started_parsing_mdat?: started_parsing_mdat?} |> update_fsm_state()
+    partial = if state.fsm_state in [:skip_mdat, :go_back_to_mdat], do: <<>>, else: rest
+
+    state = %{state | partial: partial}
 
-    if can_read_data_box?(state) do
-      handle_can_read_mdat_box(ctx, state)
+    cond do
+      state.fsm_state == :mdat_reading ->
+        handle_can_read_mdat_box(ctx, state)
+
+      state.optimize_for_non_fast_start? ->
+        state =
+          if state.fsm_state == :skip_mdat,
+            do: %{state | mdat_beginning: state.boxes_size, mdat_size: maybe_header.content_size},
+            else: state
+
+        handle_non_fast_start_optimization(state)
+
+      true ->
+        {[demand: :input], state}
+    end
+  end
+
+  defp update_fsm_state(state) do
```

Comment:
as we discussed, add the context argument

Reply:
done